### PR TITLE
DRAFT: E2E ProcessBlockPoS Testing

### PR DESCRIPTION
**Pull request by lazynina**

### CODE REVIEW

**Comment by lazynina** [comment]
Current dependencies on/for this PR:
* feature/pos-fast-hotstuff-consensus
  * **PR #742** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/742" target="_blank">Graphite</a> 
    * **PR #746** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/746" target="_blank">Graphite</a> 
      * **PR #747** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/747" target="_blank">Graphite</a> 
        * **PR #748** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/748" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/deso-protocol/core/748?utm_source=stack-comment).

File: lib/pos_blockchain_test.go [hunk 1]
```diff
@@ -1471,7 +1472,139 @@ func _verifyCommitRuleHelper(testMeta *TestMeta, committedBlocks []*BlockHash, u
...
+func TestProcessBlockPoS(t *testing.T) {
+	testMeta := NewTestPoSBlockchainWithValidators(t)
+
+	// Create a bad block and try to process it.
+	dummyBlock := _generateDummyBlock(testMeta, 12, 12, 887)
+	success, isOrphan, missingBlockHashes, err := testMeta.chain.processBlockPoS(dummyBlock, 12, true)
+	require.False(t, success)
+	require.False(t, isOrphan)
+	require.Len(t, missingBlockHashes, 0)
+	require.Error(t, err)
```
Comment (tholonious) [inline lib/pos_blockchain_test.go:1484]: Just to make the order of blocks a bit easier to eyeball, it's helpful to section off all work for each block into separate code block (squiggly brackets). Really helps readability
```suggestion
	{
		// Create a bad block and try to process it.
		dummyBlock := _generateDummyBlock(testMeta, 12, 12, 887)
		success, isOrphan, missingBlockHashes, err := testMeta.chain.processBlockPoS(dummyBlock, 12, true)
		require.False(t, success)
		require.False(t, isOrphan)
		require.Len(t, missingBlockHashes, 0)
		require.Error(t, err)
	}
```

File: lib/pos_blockchain_test.go [hunk 2]
```diff
@@ -1471,7 +1472,139 @@ func _verifyCommitRuleHelper(testMeta *TestMeta, committedBlocks []*BlockHash, u
 	}
 }
 
-func _generateBlockAndAddToBestChain(testMeta *TestMeta, blockHeight uint64, view uint64, seed int64) *MsgDeSoBlock {
+func TestProcessBlockPoS(t *testing.T) {
```
Comment (tholonious) [inline lib/pos_blockchain_test.go:1475]:
```suggestion
// Test the following series of blocks to make sure that ProcessBlockPoS properly handles all cases as expected during the steady state
// 1. Process a bad block. The block could be bad for any reason, we don't really care the reason, we just want to // see it get rejected.
// 2. Process three good blocks in a row, which tests the commit rule
// 3. Process a timeout block that reorgs the previous tip
// 4. Process a regular block that reorgs from the previous tip
// 5. Process an orphan, which tests the block's storage and the return value of missingBlockHashes
func TestProcessBlockPoS(t *testing.T) {
```

File: lib/pos_blockchain_test.go [hunk 3]
```diff
@@ -1471,7 +1472,139 @@ func _verifyCommitRuleHelper(testMeta *TestMeta, committedBlocks []*BlockHash, u
...
+	require.True(t, success)
+	require.False(t, isOrphan)
+	require.Len(t, missingBlockHashes, 0)
+	require.NoError(t, err)
+
+	// Okay now we can check the best chain.
+	// We expect the block to be uncommitted.
+	blockHash1, err := realBlock.Hash()
+	require.NoError(t, err)
+	_verifyCommitRuleHelper(testMeta, []*BlockHash{}, []*BlockHash{blockHash1}, nil)
```
Comment (tholonious) [inline lib/pos_blockchain_test.go:1498]: nit: similar comment as above
```suggestion
	// Test a good block with a vote QC
	{
		realBlock = _generateRealBlock(testMeta, 12, 12, 889, testMeta.chain.GetBestChainTip().Hash)
		success, isOrphan, missingBlockHashes, err = testMeta.chain.processBlockPoS(realBlock, 12, true)
		require.True(t, success)
		require.False(t, isOrphan)
		require.Len(t, missingBlockHashes, 0)
		require.NoError(t, err)

		// Okay now we can check the best chain.
		// We expect the block to be uncommitted.
		blockHash1, err := realBlock.Hash()
		require.NoError(t, err)
		_verifyCommitRuleHelper(testMeta, []*BlockHash{}, []*BlockHash{blockHash1}, nil)
	}
```

**Comment by lazynina** [comment]
closed in favor of #762

**Pull request closed without merging**